go.mod

```diff
@@ -140,3 +141,6 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 	tags.cncf.io/container-device-interface/specs-go v0.7.0 // indirect
 )
+
+// Use the relative local source of the github.com/containerd/containerd/api to build
+replace github.com/containerd/containerd/api => ./api
```

Comment:
Alternatively, configure a [Go Workspace](https://go.dev/ref/mod#workspaces) and commit the `go.work` file. I suspect it'd be easier to remember to delete a file in the release branches than to edit one.

> Would it also be worth having a CI job to do some minimal testing for the actual version specified in go.mod (once there are releases?) because that's the version "others" will see as being used.

With testing only on release branches, only tagged releases of containerd would be (somewhat) guaranteed to be importable as a module. Downstream consumers of the containerd client might not be able to try out the `HEAD` version of the module, e.g. for early integration/acceptance testing with downstream projects, unless the minimum api dependency version is always kept up to date with the actual source code requirements. I think that CI, including PR builds, should _always_ build and run tests using the minimum versions of the dependencies a downstream importer might consume. (Read: CI should test containerd PRs against the minimum version of the api module `require`d in containerd's `go.mod`.) This would enforce a workflow where api module changes would always have to be committed, and the api dependency bumped, before the changes could be used from the containerd module. I think that enforcing such a strict workflow _would be to the benefit of containerd development_ as it ensures that the `github.com/containerd/containerd/v2@main` module is always importable.

With a committed `go.work`, CI could run builds/tests with `GOWORK=off` in the environment so the file is ignored. Alternatively, `.gitignore` the `go.work` path and provide a script/template for containerd developers to quickly get a local workspace configured. You never have to remember to delete the `go.work` file in release branches if it is never committed in the first place. The `go.mod` equivalent to either approach is more awkward and error-prone: you can't `.gitignore` a portion of a file, and CI would have to `go mod edit` the replace directive away to test PRs how external importers would see them.

Reply:
I like the idea of go.work but found it awkward to use when I first tried it out. If it is not committed, how would you be able to have a pull request for an api change and its implementation in the same PR?

Reply:
One idea, untested: `git push -u my-fork my-pr-branch && go get github.com/containerd/containerd/api@"$(git show --pretty='%H')"`. Or perhaps some offline `go mod edit` command could also work. As commits pushed to a fork of a GitHub repo are reachable from the main repo by commit hash, any commit pushed to a PR branch on GitHub is also a valid module pseudo-version.

Reply:
It probably isn't an issue for the replace to just come and go, after we tag the version and update, we remove the replace. When we add a change and the replace isn't there, we re-add it until it gets updated back to a tag. We shouldn't need it as a part of any tagged and can add that to the release checklist.